Automattic/wp-calypso
# Gutenberg: Hide opt-in dialogs if user cannot switch to the block editor

#### Changes proposed in this Pull Request

We now check if the current site can switch to the block editor Gutenberg before enabling the opt-in prompts.

This will prevent users from seeing the "Switch to the block editor" dialog if their sites are unable to load Gutenberg in WP.com due to the following conditions:
- Site is running a version WP lower than 5.0 and the Gutenberg plugin is not active. This is determined by the `opt_in` field of the Gutenberg opt-in API response (introduced in D31164-code).
- The Classic Editor plugin is active (effectively opting out of Gutenberg). This is determined by the `opt_in` field of the Gutenberg opt-in API response (introduced in D31164-code).
- Ineligible JP/AT sites for Gutenframe when using the desktop app.
- VIP sites.

This PR also cleans up the feature flags `gutenberg/redirect` and `gutenberg/opt-in` since they are not needed anymore.

#### Testing instructions

This needs to be tested on the desktop app, so make sure you have the dev environment [installed and running](https://github.com/Automattic/wp-desktop/blob/468ecd4d68cdafbc2935facfcd15953f46db98ce/docs/install.md) (more info about how to set up the secrets at paaHJt-7y-p2).

- Apply D31164-code and sandbox the API.
- Make sure you don't have any active denied plugin for Gutenframe in your AT test site (see PCYsg-lOu-p2 for information about denied plugins).
- Make sure the Classic Editor plugin is not activated in your AT test site.
- In the desktop repo, point the calypso submodule to this branch: `cd calypso && git fetch && git checkout fix/35039-gutenberg-hide-opt-in-dialogs && cd ..`.
- Generate  a new build: `make build`.
- Open the built app in the `release/<YOUR_OS>` folder (i.e. `release/mac/WordPress.com.app`).
- Switch to your AT test site.
- Start creating a post.
- Make sure the block editor is loaded in the desktop app.
- Switch to the classic editor from either the inline help or the more menu:
<img width="150" alt="Screen Shot 2019-07-24 at 12 05 56" src="https://user-images.githubusercontent.com/1233880/61762003-7c9c4480-ae0b-11e9-9db2-15f7cd79cd1e.png"><img width="100" alt="Screen Shot 2019-07-24 at 12 06 15" src="https://user-images.githubusercontent.com/1233880/61762008-7f973500-ae0b-11e9-964c-cbc0327bf598.png">

- Make sure a warning about the post containing blocks shows up:
<img width="300" alt="Screen Shot 2019-07-24 at 12 08 48" src="https://user-images.githubusercontent.com/1233880/61762080-cf75fc00-ae0b-11e9-8da0-8ff4a6738eb3.png">

- Click on the "Use the classic editor" button.
- Switch to the block editor from either the inline help or the sidebar:
<img width="120" alt="Screen Shot 2019-07-24 at 12 10 36" src="https://user-images.githubusercontent.com/1233880/61762171-1532c480-ae0c-11e9-880a-6d5c27669d0f.png"><img width="120" alt="Screen Shot 2019-07-24 at 12 10 26" src="https://user-images.githubusercontent.com/1233880/61762168-12d06a80-ae0c-11e9-8c12-c0f9d1758bbb.png">

- Install and activate the Classic Editor plugin on your AT test site.
- Close the desktop app and open it again, so the cache is cleared and the new installed plugin is considered when determining which editor should be loaded (note that this extra step is needed due to the issue noted in #33074).
- Create a new post again.
- Verify the Calypso classic editor is loaded now in the desktop app.
- Make sure you cannot switch to the block editor now (this step can be tested in both the web and the desktop app).
- Deactivate the Classic Editor plugin.
- Install and activate a denied plugin for Gutenframe.
- Close the desktop app and open it again.
- Create a new post again.
- Verify the Calypso classic editor is loaded again in the desktop app (this step can be tested only in the desktop app).
- Make sure you cannot switch to the block editor (this step can be tested only in the desktop app).

Fixes #35039.
